### ISSUE TITLE
Stop gradient

### ISSUE THREAD

**Issue by chrislybaer** [XifengGuo/CapsNet-Keras#26]
Why is this necessary?

https://github.com/XifengGuo/CapsNet-Keras/blob/0ca571cb6c0fb465befebcb4e1fccef6efaecda3/capsulelayers.py#L152

**Comment by XifengGuo** [comment]
@chrislybaer In my understanding, the routing is just to search best connections `c` between capsules from one layer and capsules from next layer in a dynamic way. So when we have found the best connections in forward, we should only use the best connections to formulate the relationship between `inputs_hat` and `outputs`. That is, we just need to consider the gradient backpropagated through the best connections (i.e. the `b` used in the last iteration) and ignore gradients from inner iterations.    
Therefore, we use `inputs_hat_stopped = K.stop_gradient(inputs_hat)` to stop gradients from inner iterations flowing back to `inputs_hat`.

**Comment by chrislybaer** [comment]
Okay thank you, this explanation sounds intuitive!

**Comment by saralajew** [comment]
I'm not 100% percent sure if that what you are argue is right. In the paper of Hinton they described the routing algorithm as a procedure returning `v_j` after the whole iterations. If it would be the intention to use the gradient flow just over the final setting of the coupling coefficients `c`. They would propose the algorithm as returning `c` and using the `c`'s  out-of-the-box in the routing of the neural network. But that's not the case. Thus, I have some concerns that this is 'right' in the sense of the paper. Furthermore, you find implementations where the authors implement the method without gradient_stop (e.g. https://github.com/InnerPeace-Wu/CapsNet-tensorflow/blob/master/code/CapsNet.py). Moreover, in my understanding the most Pytorch implementations also not using the gradient_stop idea (but I'm not so familiar with the Pytorch coding).

From my point of view it is just a minor change and both implementations should be fine. Using gradient_stop will focus the gradient back flow to the final `c`'s and hiding the information how we built the `c`'s. So there is no information back propagation that the `c`'s where built by the values of `u`.

Another comment: In the paper they proposed that the initial values for `b` could be trained by the neural network (again, see: https://github.com/InnerPeace-Wu/CapsNet-tensorflow/blob/master/code/CapsNet.py). Which seems as one more point that the gradient_stop might be 'wrong'.

Maybe an experiment and comparison of both methods could help to clarify the initial question of @chrislybaer ?

**Comment by XifengGuo** [comment]
@saralajew I have tested both of them, stopping gradient leads to similar accuracy and better efficiency. More discussions are welcome.

**Comment by chrislybaer** [comment]
@saralajew regarding the "... In the paper they proposed that the initial values for b could be trained ...":
I wrote an email to the authors and they said they did not use this for the MNIST dataset and expect improvements for broader capsule networks (networks with more than 100 capsule types).

I also tested both and accuracy-wise there is not much difference. I personally do not really like the stop_gradient command because it makes the code a little bit harder to read and to imagine what is going on.

20 epochs without stop_gradient: val_capsnet_acc: 0.9952
20 epochs with stop_gradient:       val_capsnet_acc: 0.9950

**Comment by saralajew** [comment]
The aim of my comment was just to clarify the statement of @XifengGuo and, further, to avoid that the issue#26 provide some 'theoretical' wrong understandings :smile: 
@chrislybaer: see page 3 footnote 2: The authors state they use fixed initial `b`'s.
Moreover, it seems intuitive that the computational efficiency should be better with gradient stop (less gradient back flow).

Thanks for the comments and @chrislybaer for providing some results!

**Comment by XifengGuo** [comment]
@chrislybaer @saralajew 
I still insist my understanding explained before. But I find some problems of using gradient_stop.

I conducted an experiment with a small capsnet structure and found that removing gradient_stop performs better in terms of acc, though it took a little more time.
Furthermore, gradient_stop tends to lead to insufficient training for transformation (prediction) matrices, i.e. some capsules in the lower layer cannot predict correct capsule in the higher layer. But we expect transformation matrices to be trained perfectly in order to benefit routing algorithms (Standing on giant's shoulder sees further?) 

One possible way to solve this issue is to train the capsnet with one iteration of routing for some epochs and then to train with 3 iterations of routings. The pretraining step can help the transformation matrices be well trained.
 
But since removing gradient_stop can overcome this, I decide to just abandon gradient_stop.

**Comment by ntajbakhsh** [comment]
Based on my understanding, both tensors b and c are constant, because b is initialized as tf.zeros and c is just a softmax on b. As such, both are not trainable. Also, u^hat is the output of a matrix multiplication and thus a constant tensor. Therefore, applying stop_gradient on these tensors seems unnecessary. Please correct me if I'm wrong.
